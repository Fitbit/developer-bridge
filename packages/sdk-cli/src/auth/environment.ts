import process from 'process';
// tslint:disable-next-line:import-name
import environments from './environments.json';

export default () => {
  const defaultEnv = 'production';
  const environment = (
    process.env.FITBIT_SDK_ENVIRONMENT || defaultEnv
  ) as keyof typeof environments;

  if (Object.keys(environments).indexOf(environment) === -1) {
    throw new Error(`Invalid environment specified: ${environment}`);
  }

  const config = environments[environment];

  const clientId = process.env.FITBIT_SDK_CLIENT_ID;
  const clientSecret = process.env.FITBIT_SDK_CLIENT_SECRET;

  if (clientId || clientSecret) {
    if (!clientId || !clientSecret) {
      throw new Error('Both FITBIT_SDK_CLIENT_ID and FITBIT_SDK_CLIENT_SECRET must be specified');
    }

    config.clientId = clientId;
    config.clientSecret = clientSecret;
  }

  return {
    environment,
    config,
  };
};
