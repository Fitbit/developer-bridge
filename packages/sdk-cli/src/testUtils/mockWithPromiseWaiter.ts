export default function mockWithPromiseWaiter<T>(
  mockFn: jest.Mock,
  returnValue?: any,
) {
  return new Promise<T>((resolve) => {
    mockFn.mockReset();
    mockFn.mockImplementationOnce((arg: T) => {
      resolve(arg);
      return returnValue;
    });
  });
}
