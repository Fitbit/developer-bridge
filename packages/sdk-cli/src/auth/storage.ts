import { TokenResponse } from '@openid/appauth/built/token_response';
import * as t from 'io-ts';
import keytar from 'keytar';

import environment from './environment';

const keyStoreService = 'fitbit-sdk';
const keyStoreAccount = environment().environment;

// tslint:disable-next-line:variable-name
const AuthStorage = t.interface({
  access_token: t.string,
  refresh_token: t.string,
  issued_at: t.number,
  expires_in: t.string,
});
type AuthStorage = t.TypeOf<typeof AuthStorage>;

function set(tokenResponse: TokenResponse) {
  return keytar.setPassword(
    keyStoreService,
    keyStoreAccount,
    JSON.stringify(tokenResponse.toJson()),
  );
}

async function get(): Promise<TokenResponse | null> {
  const authDataStr = await keytar.getPassword(keyStoreService, keyStoreAccount);
  if (authDataStr === null) return null;
  let authData;
  try {
    authData = JSON.parse(authDataStr);
  } catch (ex) {
    await clear();
    return null;
  }
  if (!AuthStorage.is(authData)) {
    await clear();
    return null;
  }
  return new TokenResponse(authData);
}

function clear() {
  return keytar.deletePassword(keyStoreService, keyStoreAccount);
}

export default {
  set,
  get,
  clear,
};
