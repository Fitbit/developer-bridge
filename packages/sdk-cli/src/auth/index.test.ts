import { TokenResponse, TokenResponseJson } from '@openid/appauth/built/token_response';
import nock from 'nock';
import opener from 'opener';
import url from 'url';

import * as auth from '.';
import environment from './environment';
import fetch from '../fetch';
import mockWithPromiseWaiter from '../testUtils/mockWithPromiseWaiter';
import storage from './storage';
import { NodeCrypto } from '@openid/appauth/built/node_support';

jest.mock('./storage');
jest.mock('opener');

const mockUsername = 'test@example.com';
const mockPassword = 'p@ssw0rd';

const mockTime = 1529366247520;

const mockTokenResponseData: TokenResponseJson = {
  access_token: 'newAccess',
  refresh_token: 'newRefresh',
  token_type: 'bearer',
  expires_in: 60,
};

const mockTokenStorageData = {
  ...mockTokenResponseData,
  issued_at: mockTime,
};

let getAuthStorageSpy: jest.SpyInstance;
let setAuthStorageSpy: jest.SpyInstance;
let clearAuthStorageSpy: jest.SpyInstance;
let callbackURLPromise: Promise<string>;

function mockTokenResponse(code = 200, body: {} = mockTokenResponseData) {
  return nock(environment().config.apiUrl)
    .post('/oauth2/token')
    .reply(code, body);
}

function mockTokenResponseError(status = 500, errorType = 'internal_error') {
  mockTokenResponse(status, {
    errors: [
      {
        errorType,
        message: 'Something went wrong!',
      },
    ],
    success: false,
  });
}

function mockRevokeResponse(code = 200) {
  return nock(environment().config.apiUrl)
    .post('/oauth2/revoke')
    .reply(code);
}

function mockStoredAuthData(data = mockTokenStorageData) {
  getAuthStorageSpy.mockReset();
  getAuthStorageSpy.mockResolvedValueOnce(
    new TokenResponse(data),
  );
}

function mockEmptyAuthData() {
  getAuthStorageSpy.mockReset();
  getAuthStorageSpy.mockResolvedValueOnce(null);
}

function getCallbackURLPromise() {
  const openerPromise = mockWithPromiseWaiter<string>(opener as jest.Mock);
  return openerPromise.then((authorizeUrlStr) => {
    const authorizeUrl = new url.URL(authorizeUrlStr);
    const redirectUri = authorizeUrl.searchParams.get('redirect_uri') as string;
    return redirectUri;
  });
}

beforeEach(() => {
  nock.disableNetConnect();
  nock.enableNetConnect('127.0.0.1');

  getAuthStorageSpy = jest.spyOn(storage, 'get');
  setAuthStorageSpy = jest.spyOn(storage, 'set');
  clearAuthStorageSpy = jest.spyOn(storage, 'clear');
  jest.spyOn(Date.prototype, 'getTime').mockReturnValue(mockTime * 1000);
});

afterEach(() => nock.cleanAll());

describe('getAccessToken()', () => {
  const authStorageTemplate = {
    ...mockTokenResponseData,
    access_token: 'access',
    refresh_token: 'refresh',
  };

  it('returns null if storage is empty', () => {
    mockEmptyAuthData();
    return expect(auth.getAccessToken()).resolves.toBe(null);
  });

  it('returns access token if storage contains valid, unexpired tokens', () => {
    mockStoredAuthData();
    return expect(auth.getAccessToken()).resolves.toBe('newAccess');
  });

  describe('if the token is expired', () => {
    beforeEach(() => {
      mockStoredAuthData({
        ...authStorageTemplate,
        issued_at: mockTime - 300,
      });
    });

    it('refreshes token and returns if required', () => {
      mockTokenResponse();
      return expect(auth.getAccessToken()).resolves.toBe(mockTokenResponseData.access_token);
    });

    it('stores newly refreshed tokens', async () => {
      mockTokenResponse();
      await auth.getAccessToken();
      expect(setAuthStorageSpy).toBeCalledWith(new TokenResponse(mockTokenResponseData));
    });

    it('clears auth storage if refresh fails', async () => {
      mockTokenResponseError();
      await expect(auth.getAccessToken()).rejects.toBeDefined();
      expect(clearAuthStorageSpy).toBeCalled();
    });
  });
});

describe('loginAuthCodeFlow()', () => {
  beforeEach(() => {
    jest.spyOn(NodeCrypto.prototype, 'generateRandom')
      .mockReturnValue('fixedstate');
    callbackURLPromise = getCallbackURLPromise();
  });

  afterEach(async () => {
    // Call the authorize callback just in case the test didn't
    // so that the server doesn't remain running after the test completes.
    try {
      await fetch(await callbackURLPromise);
    } catch (ex) {}
  });

  it('retrieves and stores an auth token', async () => {
    const mockTokenEndpoint = mockTokenResponse();
    const loginPromise = auth.loginAuthCodeFlow();
    await fetch(`${await callbackURLPromise}?state=fixedstate&code=__valid_code__`);
    await loginPromise;

    expect(mockTokenEndpoint.isDone()).toBe(true);
    expect(setAuthStorageSpy).toBeCalledWith(new TokenResponse(mockTokenResponseData));
  });

  describe.each([
    'state=_invalid_&code=__valid_code__',
    'state=fixedstate&error=internal_error',
    'state=fixedstate&error=internal_error&error_description=Something+went+very_wrong',
  ])('given authorization callback parameters %s', (queryParams) => {
    it('rejects', async () => {
      const loginExpectPromise = expect(auth.loginAuthCodeFlow()).rejects
        .toThrowErrorMatchingSnapshot();
      await fetch(`${await callbackURLPromise}?${queryParams}`);
      return loginExpectPromise;
    });
  });

  it('rejects if token response returns a 500 status code', async () => {
    mockTokenResponseError();
    const loginPromise = auth.loginAuthCodeFlow();
    await fetch(`${await callbackURLPromise}?state=fixedstate&code=__valid_code__`);
    return expect(loginPromise).rejects.toThrowErrorMatchingSnapshot();
  });

  it('rejects if token response is empty', async () => {
    mockTokenResponse(200, {});
    const loginPromise = auth.loginAuthCodeFlow();
    await fetch(`${await callbackURLPromise}?state=fixedstate&code=__valid_code__`);
    return expect(loginPromise).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('loginResourceOwnerFlow()', () => {
  it('retrieves and stores an auth token', async () => {
    const mockTokenEndpoint = mockTokenResponse();
    const loginPromise = auth.loginResourceOwnerFlow(mockUsername, mockPassword);
    await expect(loginPromise).resolves.toBeUndefined();

    expect(mockTokenEndpoint.isDone()).toBe(true);
    expect(setAuthStorageSpy).toBeCalledWith(new TokenResponse(mockTokenResponseData));
  });

  it('rejects if token response returns a 500 status code', async () => {
    mockTokenResponseError();
    const loginPromise = auth.loginResourceOwnerFlow(mockUsername, mockPassword);
    return expect(loginPromise).rejects.toThrowErrorMatchingSnapshot();
  });

  it('rejects if token response is empty', async () => {
    mockTokenResponse(200, {});
    const loginPromise = auth.loginResourceOwnerFlow(mockUsername, mockPassword);
    return expect(loginPromise).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('logout()', () => {
  describe('when a user is logged in', () => {
    beforeEach(() => {
      mockStoredAuthData();
      clearAuthStorageSpy.mockReturnValueOnce(Promise.resolve());
    });

    it('deletes stored keychain data', async () => {
      mockRevokeResponse();
      await auth.logout();
      expect(clearAuthStorageSpy).toBeCalled();
    });

    it('revokes the access token', async () => {
      const revokeEndpointMock = mockRevokeResponse();
      await auth.logout();
      expect(revokeEndpointMock.isDone()).toBe(true);
    });
  });

  it('is a no-op if a user is not logged in', async () => {
    return expect(auth.logout()).resolves.toBeUndefined();
  });
});
