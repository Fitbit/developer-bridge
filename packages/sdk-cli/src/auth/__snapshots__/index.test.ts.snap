// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginAuthCodeFlow() given authorization callback parameters state=_invalid_&code=__valid_code__ rejects 1`] = `"Authorization error: Mismatched state"`;

exports[`loginAuthCodeFlow() given authorization callback parameters state=fixedstate&error=internal_error rejects 1`] = `"Authorization error: internal_error"`;

exports[`loginAuthCodeFlow() given authorization callback parameters state=fixedstate&error=internal_error&error_description=Something+went+very_wrong rejects 1`] = `"Authorization error: internal_error: Something went very_wrong"`;

exports[`loginAuthCodeFlow() rejects if token response is empty 1`] = `"Unexpected response format for status 200: {}"`;

exports[`loginAuthCodeFlow() rejects if token response returns a 500 status code 1`] = `"internal_error"`;

exports[`loginResourceOwnerFlow() rejects if token response is empty 1`] = `"Unexpected response format for status 200: {}"`;

exports[`loginResourceOwnerFlow() rejects if token response returns a 500 status code 1`] = `"internal_error"`;
